### PR TITLE
zhengyu/create_box

### PR DESCRIPTION
**Pull request by ZhengyuDiao**
create_box api in tensor-based module

---
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/5190)

### CODE REVIEW

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 4 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/5190)* status: 4 of 7 files reviewed, 13 unresolved discussions (waiting on @yxlao and @ZhengyuDiao)

___
*[`cpp/open3d/t/geometry/TriangleMesh.h` line 371 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U3r3UC7mnE96n2g62:-N4U3r3UC7mnE96n2g63:b-gxwl01) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/cpp/open3d/t/geometry/TriangleMesh.h#L371)):*
> ```objc
>     bool HasTriangleColors() const { return HasTriangleAttr("colors"); }
> 
>     /// Factory function to create a box mesh (TriangleMeshFactory.cpp)
> ```

Rewrite as:
```
Create a box triangle mesh. One vertex of the box will be placed at the origin and the box aligns with the positive x, y, and z axes.
```
___
*[`cpp/open3d/t/geometry/TriangleMeshFactory.cpp` line 35 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U2-TC4z-cSTVZvhZ7:-N4U2-TC4z-cSTVZvhZ8:b-cfbc4c) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/cpp/open3d/t/geometry/TriangleMeshFactory.cpp#L35)):*
> ```cpp
> 
> TriangleMesh TriangleMesh::CreateBox(
>         float width /* = 1.0*/,
> ```

it is not necessary to add `/* xxx */` here, consider removing them.
___
*[`cpp/open3d/t/geometry/TriangleMeshFactory.cpp` line 41 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U3GjvEdwMYJtFxSwe:-N4U3GjvEdwMYJtFxSwf:b-ng3hmr) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/cpp/open3d/t/geometry/TriangleMeshFactory.cpp#L41)):*
> ```cpp
>         core::Dtype int_dtype /* = core::Int64*/,
>         const core::Device &device) {
>     
> ```

remove extra space here.
___
*[`cpp/open3d/t/geometry/TriangleMeshFactory.cpp` line 53 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U4hW1Ev1tAHl54lt2:-N4U4hW1Ev1tAHl54lt3:b-ktocih) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/cpp/open3d/t/geometry/TriangleMeshFactory.cpp#L53)):*
> ```cpp
>     }
> 
>     utility::LogInfo("width: {}, height: {}, depth: {}", width, height, depth);
> ```

remove unnecessary prints
___
*[`cpp/open3d/t/geometry/TriangleMeshFactory.cpp` line 66 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U4nE1A1xw55CowmsW:-N4U4nE1A1xw55CowmsX:b-9uj3gi) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/cpp/open3d/t/geometry/TriangleMeshFactory.cpp#L66)):*
> ```cpp
> 
>     // Vertices.
>     core::Tensor vertex_positions({8, 3}, float_dtype);
> ```

You can just do:
```
core::Tensor vertex_positions;
```
This will create the tensor without allocation. Since we will be allocating **again** with `core::Tensor::Init`.
___
*[`cpp/open3d/t/geometry/TriangleMeshFactory.cpp` line 67 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U5YMR6aDSbhSz3nWz:-N4U5YMR6aDSbhSz3nX-:b-m8ber4) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/cpp/open3d/t/geometry/TriangleMeshFactory.cpp#L67)):*
> ```cpp
>     // Vertices.
>     core::Tensor vertex_positions({8, 3}, float_dtype);
>     if (float_dtype == core::Float32){
> ```

Since these creation functions are small, we can afford to make copies: always create with `Float64` and then convert to `Float32` if necessary. You can even fuse the sanity check here. This will simplify the code.

```cpp
core::Tensor vertex_positions = core::Tensor::Init<float>{xxx};
if (float_dtype == core::Float32) {
    vertex_positions = vertex_positions.To(core::Float32);
} else if (float_dtype != core::Float64) {
    Print error.
}
```
___
*[`cpp/open3d/t/geometry/TriangleMeshFactory.cpp` line 80 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U6O8x6iNIGbsEzdZm:-N4U6O8x6iNIGbsEzdZn:bs8jjxz) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/cpp/open3d/t/geometry/TriangleMeshFactory.cpp#L80)):*
> ```cpp
>     }
>     
>     // Triangles.
> ```

Same comment as above. 
___
*[`cpp/pybind/t/geometry/trianglemesh.cpp` line 229 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U6ZLf27TqHs3DzG1p:-N4U6ZLf27TqHs3DzG1q:b-yl0fbk) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/cpp/pybind/t/geometry/trianglemesh.cpp#L229)):*
> ```cpp
>     // Triangle Mesh's creation APIs.
>     triangle_mesh.def_static("create_box", &TriangleMesh::CreateBox,
>                             "Factory function to create a box. The left bottom "
> ```

update the docstring to be the same as c++
___
*[`cpp/tests/t/geometry/TriangleMesh.cpp` line 458 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U6dFu3HYJow-o9snW:-N4U6dFu3HYJow-o9snX:b-45yi1q) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/cpp/tests/t/geometry/TriangleMesh.cpp#L458)):*
> ```cpp
> 
> TEST_P(TriangleMeshPermuteDevices, CreateBox){
>     // to do
> ```

Add two tests:

1. Create with default parameters, and check the value of the triangle mesh.
2. Create with custom parameters, and check the value of the triangle mesh.
___
*[`python/test/t/geometry/test_trianglemesh.py` line 41 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U3aCk0BnFoViVpHgf:-N4U3aCk0BnFoViVpHgg:b-a61qiz) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/python/test/t/geometry/test_trianglemesh.py#L41)):*
> ```Python
> 
> ```

apply style: http://www.open3d.org/docs/latest/contribute/styleguide.html
___
*[`python/test/t/geometry/test_trianglemesh.py` line 43 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U3dVC-QVj1keI8lX9:-N4U3dVC-QVj1keI8lXA:bx1zxcs) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/python/test/t/geometry/test_trianglemesh.py#L43)):*
> ```Python
> 
> def test_create_box():
>     
> ```

remove empty space
___
*[`python/test/t/geometry/test_trianglemesh.py` line 46 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U6rx68N43eR3Ny7al:-N4U6rx68N43eR3Ny7am:b-4bygje) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/python/test/t/geometry/test_trianglemesh.py#L46)):*
> ```Python
>     # Test legacy
>     # box = o3d.geometry.TriangleMesh.create_box()
>     # box.compute_vertex_normals()
> ```

remove commented code
___
*[`python/test/t/geometry/test_trianglemesh.py` line 50 at r1](https://reviewable.io/reviews/isl-org/open3d/5190#-N4U6tRN8RRmLrrhebTH:-N4U6tRN8RRmLrrhebTI:b-45yi1q) ([raw file](https://github.com/isl-org/open3d/blob/954ccdd5bd0db4a5fdf07fdd637cc43e7042dbcd/python/test/t/geometry/test_trianglemesh.py#L50)):*
> ```Python
> 
> 
>     # To implement
> ```

Add two tests:

1. Create with default parameters, and check the value of the triangle mesh.
2. Create with custom parameters, and check the value of the triangle mesh.

File: cpp/open3d/t/geometry/TriangleMesh.h
Comment (reyanshsolis) [inline cpp/open3d/t/geometry/TriangleMesh.h:385]: use `Float64` as default

Comment (yxlao) [reply cpp/open3d/t/geometry/TriangleMesh.h:385]: we have been using float32 as default for tensor-based triangle mesh, e.g. `FromLegacy()` creates floa32 mesh by default

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/5190)* status: all files reviewed, 16 unresolved discussions (waiting on @reyanshsolis and @ZhengyuDiao)

___
*[`cpp/pybind/t/geometry/trianglemesh.cpp` line 233 at r3](https://reviewable.io/reviews/isl-org/open3d/5190#-N4jeNgj8C_qf5NgtQTY:-N4jeNgj8C_qf5NgtQTZ:b-v848vb) ([raw file](https://github.com/isl-org/open3d/blob/29e14c38d548af353e24768371e5a6925435105f/cpp/pybind/t/geometry/trianglemesh.cpp#L233)):*
> ```cpp
>                              "width"_a = 1.0,
>                              "height"_a = 1.0, "depth"_a = 1.0,
>                              "vertex_dtype"_a = core::Float64,
> ```

`float_dtype`,` int_dtype`
use float32 and int64 as default
___
*[`cpp/tests/t/geometry/TriangleMesh.cpp` line 459 at r3](https://reviewable.io/reviews/isl-org/open3d/5190#-N4jcgNo1NznvW8bmt5Z:-N4jcgNo1NznvW8bmt5_:b-y6q4y) ([raw file](https://github.com/isl-org/open3d/blob/29e14c38d548af353e24768371e5a6925435105f/cpp/tests/t/geometry/TriangleMesh.cpp#L459)):*
> ```cpp
>     core::Device device = GetParam();
> 
>     // test with default parameters
> ```

nit: `// Test xxxx .`
- end with `.`
- capitalilze first character
___
*[`cpp/tests/t/geometry/TriangleMesh.cpp` line 492 at r3](https://reviewable.io/reviews/isl-org/open3d/5190#-N4jcpp275DdsKHEjl_1:-N4jcpp275DdsKHEjl_2:bnj7hl0) ([raw file](https://github.com/isl-org/open3d/blob/29e14c38d548af353e24768371e5a6925435105f/cpp/tests/t/geometry/TriangleMesh.cpp#L492)):*
> ```cpp
>     EXPECT_TRUE(box_default.GetTriangleIndices().AllClose(triangle_indices));
> 
>     // test with custom parameters
> ```

same as above
___
*[`cpp/tests/t/geometry/TriangleMesh.cpp` line 494 at r3](https://reviewable.io/reviews/isl-org/open3d/5190#-N4je7Yj1Vz0Dz0ZFNVe:-N4je7YkBPBppQv-kjV_:b-bvmj4z) ([raw file](https://github.com/isl-org/open3d/blob/29e14c38d548af353e24768371e5a6925435105f/cpp/tests/t/geometry/TriangleMesh.cpp#L494)):*
> ```cpp
>     // test with custom parameters
>     t::geometry::TriangleMesh box_custom =
>             t::geometry::TriangleMesh::CreateBox(2, 3, 4);
> ```

Same as python, specify float dtype, int dtype, device.
___
*[`python/test/t/geometry/test_trianglemesh.py` line 27 at r3](https://reviewable.io/reviews/isl-org/open3d/5190#-N4jdOqb2tva81ODyZdq:-N4jdOqb2tva81ODyZdr:b-6kx2tp) ([raw file](https://github.com/isl-org/open3d/blob/29e14c38d548af353e24768371e5a6925435105f/python/test/t/geometry/test_trianglemesh.py#L27)):*
> ```Python
> # ----------------------------------------------------------------------------
> 
> from turtle import width
> ```

remove turtle
___
*[`python/test/t/geometry/test_trianglemesh.py` line 42 at r3](https://reviewable.io/reviews/isl-org/open3d/5190#-N4jd9VT9jMNV23EYVLa:-N4jd9VT9jMNV23EYVLb:b-umwm0g) ([raw file](https://github.com/isl-org/open3d/blob/29e14c38d548af353e24768371e5a6925435105f/python/test/t/geometry/test_trianglemesh.py#L42)):*
> ```Python
> 
> 
> def test_create_box():
> ```

Add `@pytest.mark.parametrize("device", list_devices())` to iterate multiple devices. See `python\test\t\geometry\test_pointcloud.py`.
___
*[`python/test/t/geometry/test_trianglemesh.py` line 44 at r3](https://reviewable.io/reviews/isl-org/open3d/5190#-N4je07u7F7Y7BVL30Q3:-N4je07vB0Krrdi5Hiim:b-ebx80k) ([raw file](https://github.com/isl-org/open3d/blob/29e14c38d548af353e24768371e5a6925435105f/python/test/t/geometry/test_trianglemesh.py#L44)):*
> ```Python
> def test_create_box():
>     # test with default parameters
>     box_default = o3d.t.geometry.TriangleMesh.create_box()
> ```

 o3d.t.geometry.TriangleMesh.create_box(device=device)



_Code quote:_
```Python
 o3d.t.geometry.TriangleMesh.create_box()
```
___
*[`python/test/t/geometry/test_trianglemesh.py` line 61 at r3](https://reviewable.io/reviews/isl-org/open3d/5190#-N4jdll_1qfLx0zWexX-:-N4jdll_1qfLx0zWexX0:b-tckc49) ([raw file](https://github.com/isl-org/open3d/blob/29e14c38d548af353e24768371e5a6925435105f/python/test/t/geometry/test_trianglemesh.py#L61)):*
> ```Python
>                                             [1, 5, 7], [1, 7, 3], [2, 3, 7],
>                                             [2, 7, 6], [0, 4, 1], [1, 4, 5]]),
>                                   dtype=o3c.Dtype.Int64,
> ```


Remove `.Dtype`
___
*[`python/test/t/geometry/test_trianglemesh.py` line 68 at r3](https://reviewable.io/reviews/isl-org/open3d/5190#-N4jdWsk7Tmn5zW3wv8u:-N4jdWsl6-DI4vPf2bnT:b-6ima78) ([raw file](https://github.com/isl-org/open3d/blob/29e14c38d548af353e24768371e5a6925435105f/python/test/t/geometry/test_trianglemesh.py#L68)):*
> ```Python
> 
>     # test with custom parameters
>     box_custom = o3d.t.geometry.TriangleMesh.create_box(2, 4, 3)
> ```

`o3d.t.geometry.TriangleMesh.create_box(2, 3, 4, o3c.float64, o3c.int32, devcie)`



_Code quote:_
```Python
o3d.t.geometry.TriangleMesh.create_box(2, 4, 3)
```

File: python/test/t/geometry/test_trianglemesh.py
Comment (reyanshsolis) [inline python/test/t/geometry/test_trianglemesh.py:70]: you may remove `np.array`. A tensor can be directly initialized using a list. Same for the rest of PR.

File: python/test/t/geometry/test_trianglemesh.py
Comment (reyanshsolis) [inline python/test/t/geometry/test_trianglemesh.py:78]: we can use `o3c.float64` instead of `dtype=o3c.Float64`. Same for rest of the PR.

File: python/test/t/geometry/test_trianglemesh.py
Comment (reyanshsolis) [inline python/test/t/geometry/test_trianglemesh.py:67]: add a `.` after all comments in this PR.

File: python/test/t/geometry/test_trianglemesh.py
Comment (reyanshsolis) [inline python/test/t/geometry/test_trianglemesh.py:27]: it seems we are not using it anywhere. Remove the import if that's the case.

File: cpp/tests/t/geometry/TriangleMesh.cpp
Comment (reyanshsolis) [inline cpp/tests/t/geometry/TriangleMesh.cpp:459]: `.`

**Review by reyanshsolis — CHANGES_REQUESTED** [review]

**Review by yxlao — APPROVED** [review]
lgtm after ci

**Comment by yxlao** [comment]
Wait for #5248 to pass before merging.

**Pull request closed without merging**
